rclpy/src/rclpy/_rclpy_pycapsule.c

```diff
@@ -0,0 +1,161 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Python.h>
+
+/// Get the name of a pycapsule.
+/**
+ * Raises TypeError if the argument is not a pycapsule
+ *
+ * \param[in] pycapsule a pycapsule
+ * \return Name or None if the capsule has no name.
+ */
+static PyObject *
+rclpy_pycapsule_name(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pycapsule;
+  if (!PyArg_ParseTuple(args, "O", &pycapsule)) {
+    return NULL;
+  }
+
+  const char * name = PyCapsule_GetName(pycapsule);
+
+  if (PyErr_Occurred()) {
+    return NULL;
+  }
+
+  if (NULL == name) {
+    Py_RETURN_NONE;
+  }
+  return PyUnicode_FromString(name);
+}
+
+/// Get the address held by a pycapsule.
+/**
+ * Raises TypeError if the argument is not a pycapsule
+ *
+ * \param[in] pycapsule a pycapsule
+ * \return integer with the address held by the pycapsule.
+ */
+static PyObject *
+rclpy_pycapsule_pointer(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pycapsule;
+  if (!PyArg_ParseTuple(args, "O", &pycapsule)) {
+    return NULL;
+  }
+
+  const char * name = PyCapsule_GetName(pycapsule);
+
+  if (PyErr_Occurred()) {
+    return NULL;
+  }
+
+  void * pointer = PyCapsule_GetPointer(pycapsule, name);
+
+  if (PyErr_Occurred()) {
+    return NULL;
+  }
+
+  return PyLong_FromUnsignedLongLong((uint64_t)pointer);
```

Comment:
This assumes the size of a pointer, which will probably "just work", but maybe there's a better way to handle this?

Reply:
Added `_Static_assert` in 5e40a23. Hopefully visual studio supports this, but if not I think I can `#ifdef` and use [_STATIC_ASSERT](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/static-assert-macro?view=vs-2019)

Reply:
Did you mean better way to represent the pointer address in python, or better way to work with entities than rely on the pointer address?

If the latter, it would be nice if the rclpy wait set API said which entity was ready without having to look it up the pycapsule's address.

Reply:
Static assert is good.

Both I guess? I was thinking like using `void *` would be better, and having like a `PyPointer` like thing (not sure that exists) would be better, but maybe that's sort of what a capsule is?

Reply:
Yeah [`PyCapsule`](https://docs.python.org/3/c-api/capsule.html) is the only pointer thing I know of, but it doesn't expose the address it holds to python code.

Reply:
Aha, there is better. The builtin function `id` returns the address in memory of an object. [It uses `PyLong_FromVoidPtr()`.](https://github.com/python/cpython/blob/d246a6766b9d8cc625112906299c4cb019944300/Python/bltinmodule.c#L1206-L1211)

Reply:
69399f5